### PR TITLE
Add support for modifying existing RDS service instances

### PR DESCRIPTION
**Pull request by ccostino**
This changeset adds support for existing service instances in cloud.gov.  While this changeset adds the underlying support necessary to support this across the board, at the moment there are several caveats:

- It is only intended for RDS instances, and only PostgreSQL and MySQL ones at that
- Shared DB instances cannot be updated
- Only the instance class can be changed

More support can be added for additional updates and more services over time.

TODO:

- [x] Clean up any cruft
- [x] Get new tests working
- [x] Add more RDS tests as needed
- [x] Enable updates for dedicated PostgreSQL and MySQL plans in the catalog
- [x] Add explicit Redis instance tests to make sure updating is disabled for now
- [x] Make sure the functionality works in the dev environment
- [x] Add smoke tests for modifying RDS instances
- [x] Fix MySQL smoke tests and uncover the underlying parameter group limit issue

## Security considerations

There shouldn't be any additional security considerations in adding this functionality.  This only allows customers to change the instance class of existing RDS instances based on the choices defined in our catalog, there are no other changes that take place such modifying subnets, VPC security groups, or credentials.

## Follow up work

The custom parameter groups aren't being removed when deleting a service instance.  More details are in the [parent issue](https://github.com/cloud-gov/product/issues/1367#issuecomment-663753750).

### CODE REVIEW

**Comment by ccostino** [comment]
There are a couple of small things left to do and then of course test this in the dev environment, but the major bulk of the work is finished and this is now ready for at least a first pass review!

File: catalog-template.yml [hunk 1]
```diff
@@ -256,23 +263,24 @@ rds:
...
+      description: "Dedicated small RDS MySQL DB instance"
       metadata:
         bullets:
           - "Dedicated RDS instance"
           - "MySQL instance"
         costs:
           -
             amount:
-              usd: 0.020
+              usd: 0.041
```
Comment (bengerman13) [inline catalog-template.yml:275]: not specifically related to this pr:
is the `costs` metadata for us to track how much a thing costs us, or is it for some feature we're not using to bill customers?

File: catalog-template.yml [hunk 2]
```diff
@@ -55,6 +55,7 @@ rds:
       instanceClass: db.t3.micro
       allocatedStorage: 10
       dbType: postgres
+      plan_updateable: true
```
Comment (bengerman13) [inline catalog-template.yml:58]: nit: `updateable` -> `updatable`

File: main.go [hunk 3]
```diff
@@ -66,8 +66,12 @@ func App(settings *config.Settings, DB *gorm.DB) *martini.ClassicMartini {
 	})
 
 	// Create the service instance (cf create-service-instance)
+	// TODO:  Shouldn't this be a POST? https://github.com/go-martini/martini#routing
```
Comment (bengerman13) [inline main.go:69]: https://github.com/openservicebrokerapi/servicebroker/blob/v2.16/spec.md#provisioning

it's a PUT

File: catalog-template.yml (hunk 2 above)
Comment (ccostino) [reply catalog-template.yml:58]: Agreed with you there!  Unfortunately I can't change that, as that's how it's defined in the [Open Service Broker API](https://github.com/openservicebrokerapi/servicebroker/blob/v2.16/spec.md#updating-a-service-instance).

File: main.go (hunk 3 above)
Comment (ccostino) [reply main.go:69]: Thank you!  That's a bit incongruous with the HTTP spec (POST is used for create things), but c'est la vie. :-)

File: catalog-template.yml (hunk 1 above)
Comment (ccostino) [reply catalog-template.yml:275]: I'll read up on this more, but I understood it as how much a customer is charged for the service.  I'll continue to dig, though.

File: catalog-template.yml (hunk 2 above)
Comment (bengerman13) [reply catalog-template.yml:58]: :ohno:

File: manager.go [hunk 4]
```diff
@@ -45,9 +45,51 @@ func createInstance(req *http.Request, c *catalog.Catalog, brokerDb *gorm.DB, id
...
+		if err != nil {
+			return response.NewErrorResponse(http.StatusBadRequest, err.Error())
+		}
+	}
+	return resp
+}
+
+func modifyInstance(req *http.Request, c *catalog.Catalog, brokerDb *gorm.DB, id string, settings *config.Settings) response.Response {
+	// Extract the request information.
+	modifyRequest, resp := request.ExtractRequest(req)
```
Comment (bengerman13) [inline manager.go:59]: should `resp` be `err`, here and on 65 and 71?

File: services/rds/broker.go [hunk 5]
```diff
@@ -142,6 +143,87 @@ func (broker *rdsBroker) CreateInstance(c *catalog.Catalog, id string, createReq
...
+
+	// Fetch the new plan that has been requested.
+	newPlan, newPlanErr := c.RdsService.FetchPlan(modifyRequest.PlanID)
+	if newPlanErr != nil {
+		return newPlanErr
+	}
+
+	// We shouldn't ever arrive to this as upgrades on the shared DB adapter are
+	// not allowed or enabled, but in case we do, explicitly error out.
+	if existingInstance.Adapter == "shared" {
```
Comment (bengerman13) [inline services/rds/broker.go:177]: belt _and_ suspenders 👍

Comment (ccostino) [reply services/rds/broker.go:177]: Nothing but the best! 😄

File: manager.go (hunk 4 above)
Comment (ccostino) [reply manager.go:59]: You're right, it should, which means it also needs to be changed elsewhere... 😆

Comment (ccostino) [reply manager.go]: Fixed now!

File: catalog-template.yml (hunk 1 above)
Comment (ccostino) [reply catalog-template.yml:275]: If we need to adjust this again in the future (and I imagine we may have to for one reason or another) we can address it in a separate PR.  Right now all I did was match the current pricing listed in AWS' documentation.

File: services/rds/broker.go [hunk 6]
```diff
@@ -142,6 +143,86 @@ func (broker *rdsBroker) CreateInstance(c *catalog.Catalog, id string, createReq
...
+	}
+
+	// Update the existing instance in the broker.
+	// TODO:  Is this necessary?  Shouldn't the adapter.modifyDB call do this?
+	//		  Or, is there a distinction between the call to AWS directly
+	//		  modifying the instance and nothing more, and we still need to
+	//		  track these types of changes ourselves on the broker side?
+	existingInstance.State = status
+	existingInstance.PlanID = newPlan.ID
+	err = broker.brokerDB.Save(&existingInstance).Error
```
Comment (ccostino) [inline services/rds/broker.go:217]: I wanted to specifically call this block out as a part of review, as I wasn't sure if this was the right approach.  Do I have the right of it here and that we need to make this updated ourselves in the broker record, or did I miss something?

Comment (bengerman13) [reply services/rds/broker.go:217]: Doing it here makes sense to me, and is in line with the provision workflow.

Comment (ccostino) [reply services/rds/broker.go:217]: Thank you!  I'll remove the comment then.

**Pull request merged**